File: src/main/java/ProductArraySolution.java
import java.util.Arrays;

public class ProductArraySolution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];

        int leftProduct = 1;
        for (int i = 0; i < n; i++) {
            result[i] = leftProduct;
            leftProduct *= nums[i];
        }

        int rightProduct = 1;
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= rightProduct;
            rightProduct = rightProduct * nums[i];
        }

        return result;
    }

    static public void main(String[] args) {

        ProductArraySolution product = new ProductArraySolution();
        int[] array = {1, 2, 3, 4};
        int[] result = product.productExceptSelf(array);

        System.out.println("Result = " + Arrays.toString(result));

    }
}

File: src/main/kotlin/ProductArraySolution.kt
class ProductArraySolutionKotlin {
    fun productExceptSelf(nums: IntArray): IntArray {
        val n = nums.size
        val result = IntArray(n)

        var leftProduct = 1
        for (i in nums.indices) {
            result[i] = leftProduct
            leftProduct *= nums[i]
            println(result[i])
        }

        var rightProduct = 1
        for (i in nums.indices.reversed()) {
            result[i] *= rightProduct
            rightProduct *= nums[i]
            println(result[i])
        }

        return result
    }
}

fun main() {
    val array = intArrayOf(1,2,3,4)
    val result = ProductArraySolutionKotlin().productExceptSelf(array)

    println("Result = $result")
}